In [154]:
import numpy
from matplotlib import pyplot
%matplotlib inline
import math

J'ai pris les mêmes bibliothèques que dans l'exemple.
Méthode: on doit créer un tableau reprenant les valeurs de h et v entre t=0 et t=T(temps final où la fusée se crashe) toutes les dt=0.1s. Si j'ai bien compris la méthode d'Euler: $$\frac{dh}{dt}=v \to h_{n+1}=h_{n}+v_{n}*dt$$ Avec la petite incertitude de savoir s'il faut prendre $v_{n}$ ou $v_{n+1}$ dans la formule? Pour v ça se complique à cause de mp. Je ne sais pas si on peut simplement écrire: $$v_{n+1}=v_{n}+dt*(-g+\frac{\dot{m_{p}}}{m_{s}+m{p}}*ve - \frac{1}{2}*\rho*v*\mid{v}\mid*A*C_{D})$$
il faudrait donc calculer mp simultanément.

Je commence par écrire toutes les valeurs initiales

In [155]:
g=9.81
p=1.091
r=0.5
A=math.pi*r**2
ve=325.0
ms=50.0
CD=0.15
dt=0.1
h0=0.0
v0=0.0

on doit calculer le temps final T jusqu'auquel on doit calculer mais le problème c'est qu'il dépend du probleme donc je sais pas comment faire. au final jai mis un nombre assez grand pour pas avoir de probleme avec les indices mais il faudrait peut etre reflechir à une maniere plus maligne de faire pcq la je pense que cest pas super optimal

In [156]:
T=100.
N=int(T/dt)+1
t=numpy.arange(0.0,T,0.1)

en fait je comprend pas très bien à quoi ça sert ça => c'est une manière de créer une liste de valeurs qu'on pourra obtenir plus tard par méthode d'Euler en remplaçant les zéros de la liste (c'est l'équivalent de partir d'une liste vide et d'append les valeurs, mais ça se révèle mieux de faire comme ça, ici), sauf pour la première qu'on pose tout de suite

In [157]:
v=numpy.zeros(N)
h=numpy.zeros(N)
v[0]=v0
h[0]=h0
ti=0.0

j'ai pris comme toi v0=0

In [158]:
def mp_value(t):
    if t<=5.0:
        mp=100.0-20.0*t
    elif t>5:
        mp=0
    return mp  

In [159]:
def mp_dot(t):
    if t<=5:
        mpdot=20.0
    else:
        mpdot=0.0
    return mpdot

j'ai crée une fonction qui calcule la valeur de mp et une autre qui donne simplement $\dot{mp}$, je pense que ça rend le code après plus lisible puisqu'il ne faudra plus refaire à chaque fois le calcul à l'interieur du code

In [160]:
n=0
while h[n]>=0.0:
    n=n+1
    ti=ti+dt
    v0=v0+dt*(-g+ ((mp_dot(ti)/(ms+mp_value(ti)))*ve)- (1/2)*p*v0*abs(v0)*A*CD)
    h0=h0+v0*dt
    v[n]=v0
    h[n]=h0

ici j'ai calculé à la fois h et v et je l'ai fait avec une boucle while pour que le critère d'arrêt ce soit que la fusée a touché le sol et j'ai eu des problème avec la taille de mes liste et du coup jai du mettre T=1000 au dessus pour plus avoir de bug

In [161]:
print(n) #je fais ça pour savoir cb d'étape il a en fait fallu

706


alors je ne comprend pas trop ce qu'il se passe pcq la valeur de n change a chaque fois que je fais ctrl shift, jai deja eu 253 et puis 706 et puis 1 et puis 706... ca n'a aucun sens

In [162]:
print(h[n])

-16.118086164


In [165]:
print(h[705])

16.9984299533
